### PR TITLE
Add Timeline plugin

### PR DESCRIPTION
**Pull request by hansthen**
I closed the earlier pull request for the `timeline` functionality as I was not happy with the approach. I think a separate plugin is better. I think the difference in functionality is sufficient to warrant its own plugin.

Besides the ability to use `end` time, this plugin also does not require a array of `times` array with matching elements in a `geometry` array. Just giving each feature a `start` and `end` is simpler.

Another `timeline` specific functionality, is that it allows multiple layers to be controlled by a single `TimeLineSlider` control.

Lastly, the `Timeline` class inherits from `GeoJson`, so it can use `GeoJsonPopup` and `GeoJsonTooltip` objects.

### CODE REVIEW

**Comment by hansthen** [comment]
Btw, if you think this plugin is best served in its own repository, I am fine with that as well.

**Comment by Conengmo** [comment]
Sorry for the delay in reviewing, I will get to it shortly.

**Comment by hansthen** [comment]
pre-commit.ci autofix

File: folium/plugins/timeline.py
Comment (Conengmo) [inline folium/plugins/timeline.py:55]: this data doesn't exist

File: folium/plugins/timeline.py
Comment (Conengmo) [inline folium/plugins/timeline.py:77]: is this necessary for the example? The example should be the minimum necessary to show how the class works. Or show case some common features. This elaborate custom styling doesn't seem that relevant for learning how the arguments work.

File: folium/plugins/timeline.py
Comment (Conengmo) [inline folium/plugins/timeline.py:214]: some of this styling seems general to Leaflet. It may impact other features, so I'd rather not have it here. Better to make it apply to this plugin only, or include it in the map styling. But for the latter that's possibly dangerous, since it may break other features.

File: folium/plugins/timeline.py
Comment (Conengmo) [inline folium/plugins/timeline.py:154]: is that true? So it can't be grouped in a featuregroup?

File: folium/plugins/timeline.py
Comment (Conengmo) [inline folium/plugins/timeline.py:260]: please make sure you include all arguments in the docstring.

**Review by Conengmo — CHANGES_REQUESTED** [review]
I'd like to have this plugin in our repo, since it seems like a better approach than our current timestamped geojson plugin. But some work is needed to get it mergeable.

I added some comments based on the code alone, I didn't test it yet because the example is missing data.

We'll also need:
- an entry in our documentation
- list this plugin in the 'plugins' page in the docs. We'll have to make sure that the description of this one and of timestamped geojson show the difference between the two.
- importing both classes in folium/plugins/__init__.py

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py:77]: It was in the original example for the timeline plugin. I tried the example without colors, but the demo looked horrible.

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py:260]: Removed speed_slider as it was not used.

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py:214]: I do not know how to apply css only to a single plugin.

I am also not sure how much of a problem it is. I think Folium is typically run inside a separate iframe, so general styling will be limited to a single instance. I ran this `timeline` together with the `timestamped_geojson` in a single Jupyter notebook and there was no interference. 

The second css rule is limited to the "leaflet-timeline-controls" class, which can only be a problem if someone decides to reuse that class.

For the first css rule: the `TimelineSlider` will take the bottom left controller position, so no other controls can be positioned there for this map. To make the slider take the full width, which looks much nicer, this styling is required. This will impact other controls that take the bottom left, but I think in practice it will not be a problem.

File: requirements.txt
Comment (hansthen) [reply requirements.txt:2]: Done

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py:114]: Done

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py:129]: Done

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py:55]: Done

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py:154]: Changed.

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py]: Changed for a simpler example.

File: docs/user_guide/plugins/timeline.md
Comment (hansthen) [inline docs/user_guide/plugins/timeline.md:1]: This is temporary so the tests pass. Once the PR is accepted I will change this to point to data inside our own repository.

**Comment by hansthen** [comment]
I added documentation and tests. I also updated the documentation of timestamped_geojson to make the reader aware of the two plugins. The tests (for now) refer to an example data file (borders.json) in the `skeate` repo. This is just to make the tests pass until this PR is accepted. I will update it to point to the Folium repo once the PR is accepted.

**Comment by hansthen** [comment]
Any chance to have a look at the updated PR?

**Comment by hansthen** [comment]
@Conengmo @ocefpaf I know Folium is a volunteer project and I greatly appreciate your work on it. It is hard to make time free next to work and family obligations. Thank you both for the effort you put in the project!

I feel a bit guilty for adding to your workload by spamming you with Pull Requests. Is there any way I can help that will lessen the demands placed on you?

File: folium/plugins/timeline.py
Comment (Conengmo) [inline folium/plugins/timeline.py:37]:
```suggestion
      get_interval: JsCode, optional
```

File: folium/plugins/timeline.py
Comment (Conengmo) [inline folium/plugins/timeline.py:229]: this argument is not in the docstring

File: folium/plugins/timeline.py
Comment (Conengmo) [inline folium/plugins/timeline.py:289]: this method uses `self.data`, but that attribute doesn't exist. Maybe remove this method?

File: folium/plugins/timeline.py
Comment (Conengmo) [inline folium/plugins/timeline.py:241]: why is this an instance attribute on `self`? Seems like it is only used for `kwargs["format_output"]`. In that case I'd put it directly there, not storing it in `self` as well.

**Review by Conengmo — CHANGES_REQUESTED** [review]
Thanks for addressing my comments! Documentation looks good 👍  I think we're really close here, some minor comments and proposed steps for hosting the example data ourselves first. After that we should be able to merge this one!

**Comment by Conengmo** [comment]
> I know Folium is a volunteer project and I greatly appreciate your work on it. It is hard to make time free next to work and family obligations. Thank you both for the effort you put in the project!

Thanks for your kind words, I really appreciate it 🙏 

> I feel a bit guilty for adding to your workload by spamming you with Pull Requests. Is there any way I can help that will lessen the demands placed on you?

Appreciate the question, but I'm happy to have high quality contributions, that's not spam! I have two thoughts. First to make PRs that are easy to review and merge. Clear description, complete, and most importantly simple and concise, so no unnecessary stuff and simple code. Second to help keep Folium maintainable in the future by being critical of the tradeoff between added complexity and new features.

If you're interested, you can also become more involved with maintaining Folium. Helping triage issues would be very useful. And also help review PRs other than your own. Are you interested in something like that?

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py:231]: Done.

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py]: Oops.

File: folium/plugins/timeline.py
Comment (hansthen) [reply folium/plugins/timeline.py]: Done.

**Comment by hansthen** [comment]
> If you're interested, you can also become more involved with maintaining Folium. Helping triage issues would be very useful. And also help review PRs other than your own. Are you interested in something like that?

I can help with both. I did start with the issue list to see if there were things I could work on or reproduce. For code reviews, how would you prefer to start? Will you use the github "request code review" feature to assign PRs to me?

**Comment by Conengmo** [comment]
> I can help with both.

That's great! I added you as a collaborator on Folium. You can now also close issues or apply labels. And you can merge PRs as well. Please use this responsibly: not merging your own code without review, and for now to start get a second opinion before merging anything consequential.

> For code reviews, how would you prefer to start? Will you use the github "request code review" feature to assign PRs to me?

Yes, but you can also take initiative to review any issue or PR that has your interest and that you feel you could help with.

**Comment by Conengmo** [comment]
@ocefpaf @martinfleis wanted to let you know as well I invited @hansthen to become a collaborator, see https://github.com/python-visualization/folium/settings/access.

File: folium/plugins/timeline.py
Comment (Conengmo) [reply folium/plugins/timeline.py:37]: You marked this as resolved, but looks like the docstring didn't change? `get_interval` is an optional parameter isn't it?

**Review by Conengmo — CHANGES_REQUESTED** [review]
Looks like `examples/data/borders.json` is still in this PR.

Nearly there!

**Comment by hansthen** [comment]
> Looks like `examples/data/borders.json` is still in this PR.
> 
> Nearly there!

I used the smaller copy from the examples (also updated in the tests).

**Review by Conengmo — APPROVED** [review]
Great, thanks!

**Pull request merged**
